mesa/datacollection.py

```diff
@@ -180,11 +178,11 @@ def _record_agents(self, model):
         rep_funcs = self.agent_reporters.values()
 
         def get_reports(agent):
-            _prefix = (agent.model.schedule.steps, agent.unique_id)
+            _prefix = (agent.model.steps_, agent.unique_id)
             reports = tuple(rep(agent) for rep in rep_funcs)
             return _prefix + reports
 
-        agent_records = map(get_reports, model.schedule.agents)
+        agent_records = map(get_reports, model.agents)
```

Comment:
This could be the cause of the `test_datacollector.py` problems.

Reply:
It appears that `model.schedule.remove(agent)` doesn't remove the agent in `model.agents`.

Reply:
Do you agree with https://github.com/projectmesa/mesa/pull/1942/commits/c4ad2b9e10c2507e502cd8cf4ff3272823275783 ? It is necessary for the test_agent_lifetime to pass.

Reply:
No, unfortunately not. The other way around is okay:
- Agent removed from model --> Should be removed from schedule
But the other way around would block multiple schedulers.

Reply:
No, unfortunatly not. The other way around is okay:
- Agent removed from model --> Should be removed from schedule
But the other way around would block multiple schedulers.

Reply:
Maybe the test just need to be updated

Reply:
@rht @Corvince this is the last point need addressing. I would propose:
- Adjusting the failing test, agent lifetime now just means something else. Remove from the schedule is like “deactivated”, removed from the model is “terminated/dead”.
- We should document this clearly, with maybe a comment in the code, PR message and release notes.

Reply:
This basically means that the 2.1.5 behavior is such that it doesn't allow multiple schedulers, because the data collector is hardcoded to only 1 scheduler. I suggest to just use my solution (`agent.remove()`) for backward compatibility, and worry about being proper in Mesa 3.0.